### PR TITLE
Fix BPF map handling on upgrade when disabling the preallocation of BPF maps

### PR DESCRIPTION
**Pull request by christarazi**
See commit msgs.

### CODE REVIEW

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Review by jibi — APPROVED** [review]

**Comment by jibi** [comment]
retest-4.9

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
4.9 CI jobs were failing legitimately due to a bug that's fixed in the 3rd commit.

**Comment by jibi** [comment]
retest-netnext

**Review by joestringer — APPROVED** [review]

**Review by brb — APPROVED** [review]
Thanks!

File: pkg/ebpf/map.go [hunk 1]
```diff
@@ -67,6 +67,9 @@ func (m *Map) OpenOrCreate() error {
 		PinPath: bpf.MapPrefixPath(),
 	}
 
+	mapType := bpf.GetMapType(bpf.MapType(m.spec.Type))
+	m.spec.Flags = m.spec.Flags | bpf.GetPreAllocateMapFlags(mapType)
```
Comment (kkourt) [inline pkg/ebpf/map.go:71]: Does the `Flags` field hold other flags besides `BPF_F_NO_PREALLOC`?
Could it in the future?

**Review by kkourt — APPROVED** [review]

File: pkg/ebpf/map.go (hunk 1 above)
Comment (joestringer) [reply pkg/ebpf/map.go:71]: There are multiple possible other flags, this is just handling the "prealloc" case which has specific requirements depending on the map type (LPM cannot be preallocated; LRU for instance *must* be preallocated)

**Comment by christarazi** [comment]
Marking with `upgrade-impact` because this could cause BPF maps, specifically the CT maps, on 4.9 kernels to be recreated on downgrades, which could interrupt traffic.

**Comment by joestringer** [comment]
Marking for v1.9 backport to address #15059 . We believe this issue was always present but is now triggering CI issues due to another recent change https://github.com/cilium/cilium/pull/14853/commits/c356cd4254d6fe1fe7cbbb430a4d8d018969d0d4 .

**Comment by joestringer** [comment]
@christarazi actually that test change is also in this PR, so if we backport that first commit into v1.9 then we will likely re-introduce the same issue for v1.8 -> v1.9 testing. Maybe let's avoid backporting that first commit "test: Change preallocate BPF maps default to false".

**Pull request merged**